File: src/builtins/cd_command.nim
import os, strutils
import builtins/builtin_command
import core/session_state

type
  CdCommand* = ref object of BuiltInCommand

proc newCdCommand*(): CdCommand =
  CdCommand(name: "cd")

method run*(self: CdCommand, args: seq[string], state: SessionState) =
  if args.len == 1:
    try:
      setCurrentDir(state.homeDir)
      state.curDir = state.homeDir
    except OSError:
      echo "Failed to go home."
      state.curDir = getCurrentDir()
  else:
    case args[1]
    of "~":
      try:
        setCurrentDir(state.homeDir)
      except OSError:
        discard
      state.curDir = state.homeDir
    of "..":
      let currentDirList = state.curDir.split('/')
      if currentDirList.len > 0:
        let sectionToRemove = currentDirList[^1]
        let newDir = state.curDir.replace("/" & sectionToRemove, "")
        try:
          setCurrentDir(newDir)
          state.curDir = newDir
        except OSError:
          discard
      else:
        state.curDir = getCurrentDir()
    of "/":
      try:
        setCurrentDir(args[1])
      except OSError:
        discard
      state.curDir = getCurrentDir()
    else:
      let newDir = state.curDir & "/" & args[1]
      try:
        setCurrentDir(newDir)
        state.curDir = newDir
      except OSError:
        discard

File: src/builtins/cd_command.rs
use std::env;
use std::path::Path;
use crate::builtins::builtin_command::BuiltInCommand;
use crate::core::session_state::SessionState;

pub struct CdCommand;

impl BuiltInCommand for CdCommand {
    fn name(&self) -> &str {
        "cd"
    }

    fn run(&self, args: &[String], state: &mut SessionState) {
        if args.len() == 1 {
            // No argument: go home
            if env::set_current_dir(&state.home_dir).is_ok() {
                state.cur_dir = state.home_dir.clone();
            } else {
                eprintln!("Failed to go home.");
                state.cur_dir = env::current_dir()
                    .map(|p| p.to_string_lossy().to_string())
                    .unwrap_or_default();
            }
        } else {
            match args[1].as_str() {
                "~" => {
                    let _ = env::set_current_dir(&state.home_dir);
                    state.cur_dir = state.home_dir.clone();
                }
                ".." => {
                    if let Some(parent) = Path::new(&state.cur_dir).parent() {
                        let new_dir = parent.to_string_lossy().to_string();
                        if env::set_current_dir(&new_dir).is_ok() {
                            state.cur_dir = new_dir;
                        }
                    }
                }
                "/" => {
                    let _ = env::set_current_dir(&args[1]);
                    state.cur_dir = env::current_dir()
                        .map(|p| p.to_string_lossy().to_string())
                        .unwrap_or_default();
                }
                _ => {
                    let new_dir = format!("{}/{}", state.cur_dir, args[1]);
                    if env::set_current_dir(&new_dir).is_ok() {
                        state.cur_dir = new_dir;
                    }
                }
            }
        }
    }
}
